libs/experimental/langchain_experimental/data_anonymizer/presidio.py

```diff
@@ -89,3 +94,133 @@ def add_recognizer(self, recognizer: EntityRecognizer) -> None:
+        language: str = "en",
+        operators: Optional[Dict[str, OperatorConfig]] = None,
+    ):
+        """
+        Args:
+            analyzed_fields: List of fields to detect and then anonymize.
+                Defaults to all entities supported by Microsoft Presidio.
+            language: Language to use for analysis. Defaults to english.
+            operators: Operators to use for anonymization.
+                Operators allow for custom anonymization of detected PII.
+                Learn more:
+                https://microsoft.github.io/presidio/tutorial/10_simple_anonymization/
+        """
+        try:
+            from presidio_analyzer import AnalyzerEngine
+        except ImportError as e:
+            raise ImportError(
+                "Could not import presidio_analyzer, please install with "
+                "`pip install presidio-analyzer`. You will also need to download a "
+                "spaCy model to use the analyzer, e.g. "
+                "`python -m spacy download en_core_web_lg`."
+            ) from e
+        try:
+            from presidio_anonymizer import AnonymizerEngine
+            from presidio_anonymizer.entities import OperatorConfig
+        except ImportError as e:
+            raise ImportError(
+                "Could not import presidio_anonymizer, please install with "
+                "`pip install presidio-anonymizer`."
+            ) from e
+
+        self.analyzed_fields = (
+            analyzed_fields
+            if analyzed_fields is not None
+            else list(get_pseudoanonymizer_mapping().keys())
+        )
+        self.language = language
+        self.operators = (
+            operators
+            if operators is not None
+            else {
+                field: OperatorConfig(
+                    operator_name="custom", params={"lambda": faker_function}
+                )
+                for field, faker_function in get_pseudoanonymizer_mapping().items()
+            }
+        )
+        self._analyzer = AnalyzerEngine()
+        self._anonymizer = AnonymizerEngine()
+        self.deanonymizer_mapping: Dict[str, Dict[str, str]] = dict()
+
+    def _create_update_deanonymizer_mapping(
+        self,
+        original_text: str,
+        analyzer_results: List[RecognizerResult],
+        anonymizer_results: EngineResult,
+    ) -> None:
+        """Create a mapping from anonymized text to original text"""
+        analyzer_results = sorted(analyzer_results, key=lambda d: d.start)
+        anonymizer_results.items = sorted(
+            anonymizer_results.items, key=lambda d: d.start
+        )
+
+        new_deanonymizer_mapping: Dict[str, Dict[str, str]] = {}
+
+        for analyzed_entity, anonymized_entity in zip(
+            analyzer_results, anonymizer_results.items
+        ):
+            original_value = original_text[analyzed_entity.start : analyzed_entity.end]
+
+            if anonymized_entity.entity_type not in new_deanonymizer_mapping:
```

Comment:
what about mixing lower/upper cases? Or if instead Foo Bar, you only have in later text Foo? Or typos?
Can we somehow have a possibility to use different matching heuristics?

Reply:
There are a lot of such things and edge cases to consider - it's certainly worth spending plenty of time on it and coming up with a solution. The question is whether this is an issue for now or for later stages of development?

Reply:
But you can easily implement a way to inject strategy of matching: have a callback or component in place, with simpler implementation for now. 
This will enable a user of the API to easily extend and improve if he wants, without modification of library code. It would also simplify adding such contribution by others by already providing guide rails.